### DISCUSSION TITLE
Extension creation instructions don't allow for bundling dependencies

### DISCUSSION THREAD

**Discussion by jkjustjoshing** [directus/directus#10900] [Archive]
The instructions on [this page](https://docs.directus.io/extensions/creating-extensions/) don't work if you want to bundle a dependency with the extension.

For example, I want to create an endpoint extension that does some Markdown parsing, but the parser dependency doesn't get bundled when running `npm run build` in the extension folder.

### To Reproduce

 * Run `npm init directus-extension`, follow instructions to create extension
 * Inside the new extension folder, run `npm install --save snarkdown`
 * Inside `src/index.js` import the snarkdown dependency, and wrap the "Hello, World!" string in a call to snarkdown
 * Run `npm run build` to bundle the dependency.
 * The instructions say to copy the new `dist/` folder into my Directus instance, but looking in that folder, there's no trace of the snarkdown code. Without the snarkdown code, the extension won't successfully run.

<details>
<summary>More info</summary>

### What version of Directus are you using?

N/A

### What version of Node.js are you using?

14.18.1

### What database are you using?

N/A

### What browser are you using?

N/A

### What operating system are you using?

macOS

### How are you deploying Directus?

N/A

</details>

**Comment by azrikahar** [comment]
>  there's no trace of the snarkdown code.

Hi, can you share more info on how there's no trace of it? Perhaps your built `dist/index.js` file plus your original `src/index.js` file? I've just tested it on my end and it seems to work. 

Just to clarify, if you were looking for `import snarkdown from 'snarkdown'` or technically `require('snarkdown')` in the built file, that _should not_ happen (by default) so that the extension can be used as-is without installing dependencies on your Directus project. Here's some detail on my test for reference:

### Original `src/index.js`

Note that I added underscores to wrap "Hello, World!" so that it will be italic in markdown.

```js
import snarkdown from 'snarkdown';

export default (router) => {
	router.get('/', (req, res) => res.send(snarkdown('_Hello, World!_')));
};
```

### Built `dist/index.js`

I've placed this file in `./extensions/endpoints/snarkdown/index.js`.

```js
"use strict";var e={"":["<em>","</em>"],_:["<strong>","</strong>"],"*":["<strong>","</strong>"],"~":["<s>","</s>"],"\n":["<br />"]," ":["<br />"],"-":["<hr />"]};function n(e){return e.replace(RegExp("^"+(e.match(/^(\t| )+/)||"")[0],"gm"),"")}function r(e){return(e+"").replace(/"/g,"&quot;").replace(/</g,"&lt;").replace(/>/g,"&gt;")}function t(o,a){var c,l,s,g,u,p=/((?:^|\n+)(?:\n---+|\* \*(?: \*)+)\n)|(?:^``` *(\w*)\n([\s\S]*?)\n```$)|((?:(?:^|\n+)(?:\t|  {2,}).+)+\n*)|((?:(?:^|\n)([>*+-]|\d+\.)\s+.*)+)|(?:!\[([^\]]*?)\]\(([^)]+?)\))|(\[)|(\](?:\(([^)]+?)\))?)|(?:(?:^|\n+)([^\s].*)\n(-{3,}|={3,})(?:\n+|$))|(?:(?:^|\n+)(#{1,6})\s*(.+)(?:\n+|$))|(?:`([^`].*?)`)|(  \n\n*|\n{2,}|__|\*\*|[_*]|~~)/gm,m=[],i="",d=a||{},h=0;function $(n){var r=e[n[1]||""],t=m[m.length-1]==n;return r?r[1]?(t?m.pop():m.push(n),r[0|t]):r[0]:n}function f(){for(var e="";m.length;)e+=$(m[m.length-1]);return e}for(o=o.replace(/^\[(.+?)\]:\s*(.+)$/gm,(function(e,n,r){return d[n.toLowerCase()]=r,""})).replace(/^\n+|\n+$/g,"");s=p.exec(o);)l=o.substring(h,s.index),h=p.lastIndex,c=s[0],l.match(/[^\\](\\\\)*\\$/)||((u=s[3]||s[4])?c='<pre class="code '+(s[4]?"poetry":s[2].toLowerCase())+'"><code'+(s[2]?' class="language-'+s[2].toLowerCase()+'"':"")+">"+n(r(u).replace(/^\n+|\n+$/g,""))+"</code></pre>":(u=s[6])?(u.match(/\./)&&(s[5]=s[5].replace(/^\d+/gm,"")),g=t(n(s[5].replace(/^\s*[>*+.-]/gm,""))),">"==u?u="blockquote":(u=u.match(/\./)?"ol":"ul",g=g.replace(/^(.*)(\n|$)/gm,"<li>$1</li>")),c="<"+u+">"+g+"</"+u+">"):s[8]?c='<img src="'+r(s[8])+'" alt="'+r(s[7])+'">':s[10]?(i=i.replace("<a>",'<a href="'+r(s[11]||d[l.toLowerCase()])+'">'),c=f()+"</a>"):s[9]?c="<a>":s[12]||s[14]?c="<"+(u="h"+(s[14]?s[14].length:s[13]>"="?1:2))+">"+t(s[12]||s[15],d)+"</"+u+">":s[16]?c="<code>"+r(s[16])+"</code>":(s[17]||s[1])&&(c=$(s[17]||"--"))),i+=l,i+=c;return(i+o.substring(h)+f()).replace(/^\n+|\n+$/g,"")}module.exports=e=>{e.get("/",((e,n)=>n.send(t("_Hello, World!_"))))};
```

### Fetching that endpoint in browser

We can see that it's in italic.

![image](https://user-images.githubusercontent.com/42867097/148491883-2e5d6372-4a7c-47d2-8858-03bc6d042d87.png)

### View source of that endpoint's response in browser

We can see snarkdown successfully parsed the underscores as `<em>` tags for italic.

![image](https://user-images.githubusercontent.com/42867097/148491950-405a2b1a-69af-4621-b9dd-c67042fd5c7d.png)

---

cc @nickrum in case there's any issue in my description here

**Reply by StefanTau** [reply]
Hi,
your example is working for me. I became aware of this discussion because I also have some trouble to use a dependency. In my case I would like to use axios in one of my extensions. So if I take your example and add `import axios from 'axios';` to the endpoint I get 

>  Cannot find module 'axios'

I use an external extensions folder. I realized that when I use the default one I don't get the error. Do you have any idea how I can use axios using my external extensions folder?

**Reply by jkjustjoshing** [reply]
Thanks for your help. I realize that my issue was that I was still doing `const snarkdown = require('snarkdown')`, instead of `import snarkdown from 'snarkdown'`. I was converting a Node.js project into an extension and forgot to move from CommonJS to ES Modules.
With everything in ESM, it's working great! Thanks!

**Reply by azrikahar** [reply]
@StefanTau By external extensions folder, I assume you mean it's not in the folder of your Directus project right? You should be able to add `axios` as a devDependency in your "external" extensions project to resolve that `Cannot find module 'axios'` error if that's what appearing in your IDE.

**Reply by StefanTau** [reply]
@azrikahar Yes, I mean the extensions folder is not inside the directus project. I already have axios added to the devDependencies and my IDE is not complaining. But once I build the extension, put the index.js file in the extensions folder and start directus I get `Cannot find module 'axios'`

**Reply by nickrum** [reply]
@StefanTau This is due to `axios` being one of the dependencies that is marked as "shared between Directus and API extensions", which results in the extensions-sdk always externalizing them. There is currently no way to disable this, but we're working on a solution.

**Reply by StefanTau** [reply]
Thanks for the clarification @nickrum
